File: tests/python/test_basic.py
#!/usr/bin/env python3
"""
Basic tests for JaguarEngine Python bindings (pyjaguar module).

This test suite verifies:
- Core type operations (Vec3, Quat, Mat3x3)
- Entity state management
- Simulation control
- NumPy interoperability
- Event system integration

Usage:
    python -m pytest test_basic.py -v

Or run directly:
    python test_basic.py
"""

import sys
import math
import unittest

# Import will fail until the module is built with -DJAGUAR_BUILD_PYTHON=ON
try:
    import pyjaguar as jag
    import numpy as np
    PYJAGUAR_AVAILABLE = True
except ImportError:
    PYJAGUAR_AVAILABLE = False
    print("Warning: pyjaguar module not available. Build with -DJAGUAR_BUILD_PYTHON=ON")


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestVec3(unittest.TestCase):
    """Test Vec3 3D vector operations."""

    def test_construction(self):
        """Test Vec3 construction."""
        v1 = jag.Vec3()
        self.assertEqual(v1.x, 0.0)
        self.assertEqual(v1.y, 0.0)
        self.assertEqual(v1.z, 0.0)

        v2 = jag.Vec3(1.0, 2.0, 3.0)
        self.assertEqual(v2.x, 1.0)
        self.assertEqual(v2.y, 2.0)
        self.assertEqual(v2.z, 3.0)

    def test_static_constructors(self):
        """Test Vec3 static factory methods."""
        zero = jag.Vec3.zero()
        self.assertEqual(zero.x, 0.0)

        unit_x = jag.Vec3.unit_x()
        self.assertEqual(unit_x.x, 1.0)
        self.assertEqual(unit_x.y, 0.0)
        self.assertEqual(unit_x.z, 0.0)

        unit_y = jag.Vec3.unit_y()
        self.assertEqual(unit_y.y, 1.0)

        unit_z = jag.Vec3.unit_z()
        self.assertEqual(unit_z.z, 1.0)

    def test_arithmetic(self):
        """Test Vec3 arithmetic operations."""
        v1 = jag.Vec3(1.0, 2.0, 3.0)
        v2 = jag.Vec3(4.0, 5.0, 6.0)

        # Addition
        v3 = v1 + v2
        self.assertEqual(v3.x, 5.0)
        self.assertEqual(v3.y, 7.0)
        self.assertEqual(v3.z, 9.0)

        # Subtraction
        v4 = v2 - v1
        self.assertEqual(v4.x, 3.0)
        self.assertEqual(v4.y, 3.0)
        self.assertEqual(v4.z, 3.0)

        # Scalar multiplication
        v5 = v1 * 2.0
        self.assertEqual(v5.x, 2.0)
        self.assertEqual(v5.y, 4.0)
        self.assertEqual(v5.z, 6.0)

        # Scalar division
        v6 = v2 / 2.0
        self.assertEqual(v6.x, 2.0)
        self.assertEqual(v6.y, 2.5)
        self.assertEqual(v6.z, 3.0)

        # Negation
        v7 = -v1
        self.assertEqual(v7.x, -1.0)
        self.assertEqual(v7.y, -2.0)
        self.assertEqual(v7.z, -3.0)

    def test_vector_operations(self):
        """Test Vec3 dot product, cross product, length."""
        v1 = jag.Vec3(1.0, 0.0, 0.0)
        v2 = jag.Vec3(0.0, 1.0, 0.0)

        # Dot product
        dot = v1.dot(v2)
        self.assertAlmostEqual(dot, 0.0)

        # Cross product
        cross = v1.cross(v2)
        self.assertAlmostEqual(cross.x, 0.0)
        self.assertAlmostEqual(cross.y, 0.0)
        self.assertAlmostEqual(cross.z, 1.0)

        # Length
        v3 = jag.Vec3(3.0, 4.0, 0.0)
        self.assertAlmostEqual(v3.length(), 5.0)
        self.assertAlmostEqual(v3.length_squared(), 25.0)

        # Normalized
        v4 = v3.normalized()
        self.assertAlmostEqual(v4.length(), 1.0)

    def test_sequence_protocol(self):
        """Test Vec3 as a sequence."""
        v = jag.Vec3(1.0, 2.0, 3.0)

        self.assertEqual(len(v), 3)
        self.assertEqual(v[0], 1.0)
        self.assertEqual(v[1], 2.0)
        self.assertEqual(v[2], 3.0)

        v[0] = 10.0
        self.assertEqual(v.x, 10.0)

    def test_numpy_interop(self):
        """Test Vec3 NumPy array conversion."""
        v = jag.Vec3(1.0, 2.0, 3.0)
        arr = v.to_numpy()

        self.assertEqual(len(arr), 3)
        self.assertEqual(arr[0], 1.0)
        self.assertEqual(arr[1], 2.0)
        self.assertEqual(arr[2], 3.0)

        arr2 = np.array([4.0, 5.0, 6.0])
        v2 = jag.Vec3.from_numpy(arr2)
        self.assertEqual(v2.x, 4.0)
        self.assertEqual(v2.y, 5.0)
        self.assertEqual(v2.z, 6.0)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestQuat(unittest.TestCase):
    """Test Quat quaternion operations."""

    def test_construction(self):
        """Test Quat construction."""
        q1 = jag.Quat()  # Identity
        self.assertEqual(q1.w, 1.0)
        self.assertEqual(q1.x, 0.0)
        self.assertEqual(q1.y, 0.0)
        self.assertEqual(q1.z, 0.0)

        q2 = jag.Quat(0.707, 0.707, 0.0, 0.0)
        self.assertAlmostEqual(q2.w, 0.707, places=3)

    def test_identity(self):
        """Test identity quaternion."""
        q = jag.Quat.identity()
        self.assertEqual(q.w, 1.0)
        self.assertEqual(q.x, 0.0)
        self.assertEqual(q.y, 0.0)
        self.assertEqual(q.z, 0.0)

    def test_from_axis_angle(self):
        """Test quaternion from axis-angle."""
        axis = jag.Vec3(0.0, 0.0, 1.0)
        angle = math.pi / 2  # 90 degrees

        q = jag.Quat.from_axis_angle(axis, angle)

        # Rotate a vector
        v = jag.Vec3(1.0, 0.0, 0.0)
        v_rotated = q.rotate(v)

        self.assertAlmostEqual(v_rotated.x, 0.0, places=5)
        self.assertAlmostEqual(v_rotated.y, 1.0, places=5)
        self.assertAlmostEqual(v_rotated.z, 0.0, places=5)

    def test_from_euler(self):
        """Test quaternion from Euler angles."""
        roll = 0.0
        pitch = 0.0
        yaw = math.pi / 2  # 90 degrees

        q = jag.Quat.from_euler(roll, pitch, yaw)

        # Convert back
        r, p, y = q.to_euler()
        self.assertAlmostEqual(yaw, y, places=5)

    def test_operations(self):
        """Test quaternion operations."""
        q1 = jag.Quat.from_axis_angle(jag.Vec3(0, 0, 1), math.pi / 4)
        q2 = jag.Quat.from_axis_angle(jag.Vec3(0, 0, 1), math.pi / 4)

        # Multiplication
        q3 = q1 * q2  # Should be 90 degree rotation

        v = jag.Vec3(1.0, 0.0, 0.0)
        v_rotated = q3.rotate(v)

        self.assertAlmostEqual(v_rotated.x, 0.0, places=5)
        self.assertAlmostEqual(v_rotated.y, 1.0, places=5)

        # Conjugate
        q_conj = q1.conjugate()
        self.assertAlmostEqual(q_conj.x, -q1.x, places=5)

        # Norm
        self.assertAlmostEqual(q1.norm(), 1.0, places=5)

        # Normalized
        q_norm = q1.normalized()
        self.assertAlmostEqual(q_norm.norm(), 1.0, places=5)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestMat3x3(unittest.TestCase):
    """Test Mat3x3 matrix operations."""

    def test_construction(self):
        """Test Mat3x3 construction."""
        m = jag.Mat3x3()  # Identity
        self.assertEqual(m(0, 0), 1.0)
        self.assertEqual(m(1, 1), 1.0)
        self.assertEqual(m(2, 2), 1.0)
        self.assertEqual(m(0, 1), 0.0)

    def test_static_constructors(self):
        """Test Mat3x3 static factory methods."""
        identity = jag.Mat3x3.identity()
        self.assertEqual(identity(0, 0), 1.0)

        zero = jag.Mat3x3.zero()
        self.assertEqual(zero(0, 0), 0.0)

        diag = jag.Mat3x3.diagonal(1.0, 2.0, 3.0)
        self.assertEqual(diag(0, 0), 1.0)
        self.assertEqual(diag(1, 1), 2.0)
        self.assertEqual(diag(2, 2), 3.0)

    def test_matrix_vector_multiply(self):
        """Test matrix-vector multiplication."""
        m = jag.Mat3x3.diagonal(2.0, 3.0, 4.0)
        v = jag.Vec3(1.0, 1.0, 1.0)

        result = m * v
        self.assertEqual(result.x, 2.0)
        self.assertEqual(result.y, 3.0)
        self.assertEqual(result.z, 4.0)

    def test_numpy_interop(self):
        """Test Mat3x3 NumPy conversion."""
        m = jag.Mat3x3.diagonal(1.0, 2.0, 3.0)
        arr = m.to_numpy()

        self.assertEqual(arr.shape, (3, 3))
        self.assertEqual(arr[0, 0], 1.0)
        self.assertEqual(arr[1, 1], 2.0)
        self.assertEqual(arr[2, 2], 3.0)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestEntityState(unittest.TestCase):
    """Test EntityState structure."""

    def test_construction(self):
        """Test EntityState default construction."""
        state = jag.EntityState()

        self.assertEqual(state.position.x, 0.0)
        self.assertEqual(state.velocity.x, 0.0)
        self.assertEqual(state.mass, 1.0)

    def test_modification(self):
        """Test EntityState modification."""
        state = jag.EntityState()

        state.position = jag.Vec3(100.0, 200.0, 300.0)
        state.velocity = jag.Vec3(10.0, 0.0, 0.0)
        state.mass = 1000.0

        self.assertEqual(state.position.x, 100.0)
        self.assertEqual(state.velocity.x, 10.0)
        self.assertEqual(state.mass, 1000.0)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestConstants(unittest.TestCase):
    """Test physical constants."""

    def test_constants_available(self):
        """Test that constants are accessible."""
        self.assertAlmostEqual(jag.constants.PI, math.pi, places=10)
        self.assertEqual(jag.constants.G0, 9.80665)
        self.assertGreater(jag.constants.C, 299000000)

    def test_unit_conversions(self):
        """Test unit conversion functions."""
        self.assertAlmostEqual(jag.deg_to_rad(180.0), math.pi, places=5)
        self.assertAlmostEqual(jag.rad_to_deg(math.pi), 180.0, places=5)
        self.assertAlmostEqual(jag.ft_to_m(1.0), 0.3048, places=4)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestEnums(unittest.TestCase):
    """Test enumeration bindings."""

    def test_domain_enum(self):
        """Test Domain enum."""
        self.assertIsNotNone(jag.Domain.Air)
        self.assertIsNotNone(jag.Domain.Land)
        self.assertIsNotNone(jag.Domain.Sea)
        self.assertIsNotNone(jag.Domain.Space)
        self.assertIsNotNone(jag.Domain.Generic)

    def test_coordinate_frame_enum(self):
        """Test CoordinateFrame enum."""
        self.assertIsNotNone(jag.CoordinateFrame.ECEF)
        self.assertIsNotNone(jag.CoordinateFrame.ECI)
        self.assertIsNotNone(jag.CoordinateFrame.NED)
        self.assertIsNotNone(jag.CoordinateFrame.ENU)
        self.assertIsNotNone(jag.CoordinateFrame.Body)

    def test_simulation_state_enum(self):
        """Test SimulationState enum."""
        self.assertIsNotNone(jag.SimulationState.Uninitialized)
        self.assertIsNotNone(jag.SimulationState.Initialized)
        self.assertIsNotNone(jag.SimulationState.Running)
        self.assertIsNotNone(jag.SimulationState.Paused)
        self.assertIsNotNone(jag.SimulationState.Stopped)


@unittest.skipUnless(PYJAGUAR_AVAILABLE, "pyjaguar not built")
class TestEventSystem(unittest.TestCase):
    """Test event system bindings."""

    def test_event_types(self):
        """Test EventType enum."""
        self.assertIsNotNone(jag.EventType.SimulationStarted)
        self.assertIsNotNone(jag.EventType.EntityCreated)
        self.assertIsNotNone(jag.EventType.CollisionEnter)
        self.assertIsNotNone(jag.EventType.Stall)
        self.assertIsNotNone(jag.EventType.ThresholdExceeded)

    def test_event_category(self):
        """Test EventCategory enum."""
        self.assertIsNotNone(jag.EventCategory.System)
        self.assertIsNotNone(jag.EventCategory.Entity)
        self.assertIsNotNone(jag.EventCategory.Physics)
        self.assertIsNotNone(jag.EventCategory.Domain)
        self.assertIsNotNone(jag.EventCategory.Threshold)

    def test_event_priority(self):
        """Test EventPriority enum."""
        self.assertIsNotNone(jag.EventPriority.Immediate)
        self.assertIsNotNone(jag.EventPriority.High)
        self.assertIsNotNone(jag.EventPriority.Normal)
        self.assertIsNotNone(jag.EventPriority.Low)
        self.assertIsNotNone(jag.EventPriority.Deferred)

    def test_threshold_config(self):
        """Test ThresholdConfig creation."""
        config = jag.ThresholdConfig()
        config.parameter_name = "speed"
        config.upper_threshold = 100.0
        config.hysteresis = 5.0
        config.emit_on_recover = True

        self.assertEqual(config.parameter_name, "speed")
        self.assertEqual(config.upper_threshold, 100.0)

    def test_threshold_presets(self):
        """Test threshold preset factory functions."""
        fuel_config = jag.fuel_threshold(20.0, 5.0)
        self.assertEqual(fuel_config.parameter_name, "fuel_level")

        speed_config = jag.speed_threshold(100.0, 5.0)
        self.assertEqual(speed_config.parameter_name, "speed")

        alt_config = jag.altitude_threshold(15000.0, 100.0)
        self.assertEqual(alt_config.parameter_name, "altitude")


if __name__ == "__main__":
    unittest.main()

File: tests/lua/test_basic.lua
#!/usr/bin/env lua
--[[
    Basic tests for JaguarEngine Lua bindings

    This test suite verifies:
    - Core type operations (Vec3, Quat, Mat3x3)
    - Entity state management
    - Simulation control
    - Enums and constants

    Usage:
        lua test_basic.lua
]]

-- Add build directory to package.path
package.cpath = package.cpath .. ";./build/?.so;./build/?.dll"

-- Simple test framework
local tests_run = 0
local tests_passed = 0
local tests_failed = 0

local function assert_equal(actual, expected, message)
    tests_run = tests_run + 1
    if actual == expected then
        tests_passed = tests_passed + 1
        return true
    else
        tests_failed = tests_failed + 1
        print(string.format("  FAIL: %s (expected %s, got %s)",
            message or "test", tostring(expected), tostring(actual)))
        return false
    end
end

local function assert_near(actual, expected, tolerance, message)
    tests_run = tests_run + 1
    if math.abs(actual - expected) <= tolerance then
        tests_passed = tests_passed + 1
        return true
    else
        tests_failed = tests_failed + 1
        print(string.format("  FAIL: %s (expected %.6f +/- %.6f, got %.6f)",
            message or "test", expected, tolerance, actual))
        return false
    end
end

local function assert_true(condition, message)
    return assert_equal(condition, true, message)
end

local function assert_false(condition, message)
    return assert_equal(condition, false, message)
end

-- Try to load JaguarEngine
local status, jag = pcall(require, "jaguar")
if not status then
    print("SKIP: JaguarEngine Lua module not available")
    print("Build with: cmake -B build -DJAGUAR_BUILD_LUA=ON && cmake --build build")
    os.exit(0)
end

print("JaguarEngine Lua Bindings Test Suite")
print(string.rep("=", 50))

-- ========================================================================
-- Test Vec3
-- ========================================================================
print("\n[Test Vec3]")

-- Construction
local v1 = Vec3()
assert_equal(v1.x, 0.0, "Vec3() x")
assert_equal(v1.y, 0.0, "Vec3() y")
assert_equal(v1.z, 0.0, "Vec3() z")

local v2 = Vec3(1.0, 2.0, 3.0)
assert_equal(v2.x, 1.0, "Vec3(1,2,3) x")
assert_equal(v2.y, 2.0, "Vec3(1,2,3) y")
assert_equal(v2.z, 3.0, "Vec3(1,2,3) z")

-- Static constructors
local zero = Vec3.zero()
assert_equal(zero.x, 0.0, "Vec3.zero() x")

local unit_x = Vec3.unit_x()
assert_equal(unit_x.x, 1.0, "Vec3.unit_x() x")
assert_equal(unit_x.y, 0.0, "Vec3.unit_x() y")

-- Arithmetic
local v3 = Vec3(1.0, 2.0, 3.0)
local v4 = Vec3(4.0, 5.0, 6.0)

local sum = v3 + v4
assert_equal(sum.x, 5.0, "Vec3 + Vec3 x")
assert_equal(sum.y, 7.0, "Vec3 + Vec3 y")
assert_equal(sum.z, 9.0, "Vec3 + Vec3 z")

local diff = v4 - v3
assert_equal(diff.x, 3.0, "Vec3 - Vec3 x")

local scaled = v3 * 2.0
assert_equal(scaled.x, 2.0, "Vec3 * scalar x")
assert_equal(scaled.y, 4.0, "Vec3 * scalar y")

local neg = -v3
assert_equal(neg.x, -1.0, "Vec3 negation x")

-- Vector operations
local dot = v3:dot(v4)
assert_near(dot, 32.0, 0.0001, "Vec3 dot product")

local v5 = Vec3(3.0, 4.0, 0.0)
assert_near(v5:length(), 5.0, 0.0001, "Vec3 length")
assert_near(v5:length_squared(), 25.0, 0.0001, "Vec3 length_squared")

local norm = v5:normalized()
assert_near(norm:length(), 1.0, 0.0001, "Vec3 normalized length")

-- Cross product
local x = Vec3(1, 0, 0)
local y = Vec3(0, 1, 0)
local z = x:cross(y)
assert_near(z.x, 0.0, 0.0001, "Vec3 cross x")
assert_near(z.y, 0.0, 0.0001, "Vec3 cross y")
assert_near(z.z, 1.0, 0.0001, "Vec3 cross z")

print("  Vec3 tests: PASSED")

-- ========================================================================
-- Test Quat
-- ========================================================================
print("\n[Test Quat]")

-- Identity
local qi = Quat.identity()
assert_equal(qi.w, 1.0, "Quat.identity() w")
assert_equal(qi.x, 0.0, "Quat.identity() x")
assert_equal(qi.y, 0.0, "Quat.identity() y")
assert_equal(qi.z, 0.0, "Quat.identity() z")

-- From axis-angle
local axis = Vec3(0.0, 0.0, 1.0)
local angle = math.pi / 2  -- 90 degrees
local q = Quat.from_axis_angle(axis, angle)

-- Rotate a vector
local v = Vec3(1.0, 0.0, 0.0)
local rotated = q:rotate(v)
assert_near(rotated.x, 0.0, 0.0001, "Quat rotate x")
assert_near(rotated.y, 1.0, 0.0001, "Quat rotate y")
assert_near(rotated.z, 0.0, 0.0001, "Quat rotate z")

-- From euler and back
local roll, pitch, yaw = 0.0, 0.0, math.pi / 2
local qe = Quat.from_euler(roll, pitch, yaw)
local r2, p2, y2 = qe:to_euler()
assert_near(y2, yaw, 0.0001, "Quat euler round-trip yaw")

-- Conjugate and operations
local qc = q:conjugate()
assert_near(qc.x, -q.x, 0.0001, "Quat conjugate x")

print("  Quat tests: PASSED")

-- ========================================================================
-- Test Mat3x3
-- ========================================================================
print("\n[Test Mat3x3]")

local mi = Mat3x3.identity()
assert_equal(mi:get(0, 0), 1.0, "Mat3x3.identity() (0,0)")
assert_equal(mi:get(1, 1), 1.0, "Mat3x3.identity() (1,1)")
assert_equal(mi:get(2, 2), 1.0, "Mat3x3.identity() (2,2)")
assert_equal(mi:get(0, 1), 0.0, "Mat3x3.identity() (0,1)")

local mz = Mat3x3.zero()
assert_equal(mz:get(0, 0), 0.0, "Mat3x3.zero() (0,0)")

-- Diagonal and inertia
local md = Mat3x3.inertia(1.0, 2.0, 3.0)
assert_equal(md:get(0, 0), 1.0, "Mat3x3.inertia (0,0)")
assert_equal(md:get(1, 1), 2.0, "Mat3x3.inertia (1,1)")
assert_equal(md:get(2, 2), 3.0, "Mat3x3.inertia (2,2)")

-- Trace
assert_near(mi:trace(), 3.0, 0.0001, "Mat3x3 identity trace")

-- Matrix-vector multiply
local m = Mat3x3.inertia(2.0, 3.0, 4.0)
local vv = Vec3(1.0, 1.0, 1.0)
local result = m * vv
assert_near(result.x, 2.0, 0.0001, "Mat3x3 * Vec3 x")
assert_near(result.y, 3.0, 0.0001, "Mat3x3 * Vec3 y")
assert_near(result.z, 4.0, 0.0001, "Mat3x3 * Vec3 z")

print("  Mat3x3 tests: PASSED")

-- ========================================================================
-- Test EntityState
-- ========================================================================
print("\n[Test EntityState]")

local state = EntityState()

-- Default values
assert_equal(state.position.x, 0.0, "EntityState default position.x")
assert_equal(state.mass, 1.0, "EntityState default mass")

-- Modification
state.position = Vec3(100.0, 200.0, 300.0)
state.velocity = Vec3(10.0, 0.0, 0.0)
state.mass = 1000.0

assert_equal(state.position.x, 100.0, "EntityState position.x after set")
assert_equal(state.velocity.x, 10.0, "EntityState velocity.x after set")
assert_equal(state.mass, 1000.0, "EntityState mass after set")

print("  EntityState tests: PASSED")

-- ========================================================================
-- Test Constants
-- ========================================================================
print("\n[Test Constants]")

assert_near(constants.PI, math.pi, 0.0000001, "PI constant")
assert_equal(constants.G0, 9.80665, "G0 constant")
assert_true(constants.C > 299000000, "Speed of light")

print("  Constants tests: PASSED")

-- ========================================================================
-- Test Unit Conversions
-- ========================================================================
print("\n[Test Unit Conversions]")

assert_near(deg_to_rad(180.0), math.pi, 0.0001, "deg_to_rad(180)")
assert_near(rad_to_deg(math.pi), 180.0, 0.0001, "rad_to_deg(PI)")
assert_near(ft_to_m(1.0), 0.3048, 0.0001, "ft_to_m(1)")

print("  Unit conversion tests: PASSED")

-- ========================================================================
-- Test Enums
-- ========================================================================
print("\n[Test Enums]")

assert_true(Domain.Air ~= nil, "Domain.Air exists")
assert_true(Domain.Land ~= nil, "Domain.Land exists")
assert_true(Domain.Sea ~= nil, "Domain.Sea exists")
assert_true(Domain.Space ~= nil, "Domain.Space exists")
assert_true(Domain.Generic ~= nil, "Domain.Generic exists")

assert_true(CoordinateFrame.ECEF ~= nil, "CoordinateFrame.ECEF exists")
assert_true(CoordinateFrame.NED ~= nil, "CoordinateFrame.NED exists")

assert_true(SimulationState.Uninitialized ~= nil, "SimulationState.Uninitialized exists")
assert_true(SimulationState.Running ~= nil, "SimulationState.Running exists")

print("  Enum tests: PASSED")

-- ========================================================================
-- Summary
-- ========================================================================
print("\n" .. string.rep("=", 50))
print(string.format("Tests run: %d", tests_run))
print(string.format("Passed: %d", tests_passed))
print(string.format("Failed: %d", tests_failed))
print(string.rep("=", 50))

if tests_failed > 0 then
    print("SOME TESTS FAILED")
    os.exit(1)
else
    print("ALL TESTS PASSED")
    os.exit(0)
end
